## **Unificación de las 5 elecciones**

En este cuaderno unificamos los 5 datasets de las elecciones entre 2011 y 2019, y hacemos un curado, eliminando columnas que no son útiles, y unificando los nombres de los municipios.

Cargamos uno de los datasets, en este caso el de las elecciones de abril de 2019, y lo inspeccionamos. Tenemos que cargar unas columnas como string, que por defecto pandas asumiría como float.

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
strings = {'Sección' : 'str', 'cod_ccaa' : 'str', 'cod_prov' : 'str', 'cod_mun' : 'str', 'cod_sec' : 'str', 'Distrito' : 'str'}

In [ ]:
df_eleccion = pd.read_csv('/content/drive/MyDrive/Proyecto_KeepCoding - Propio/Data/Gen-19-Abr/gen_A19_unif_cols_prov.txt', dtype = strings)

In [ ]:
df_eleccion.head()

,Sección,cod_ccaa,cod_prov,cod_mun,cod_sec,CCAA,Provincia,Municipio,Censo_Esc,Votos_Total,...,Renta hogar 2017,Renta hogar 2015,Renta Salarios 2018,Renta Salarios 2015,Renta Pensiones 2018,Renta Pensiones 2015,Renta Desempleo 2018,Renta Desempleo 2015,dict_res,dict_res_ord
0,022019041010400101001,01,04,04001,0400101001,Andalucía,Almería,Abla,1014,768,...,20172.0,19546.0,5574.0,4833.0,3286.0,3082.0,403.0,471.0,"{'PP': 149, 'PSOE': 326, 'Cs': 131, 'UP': 44, ...","[('PSOE', 326), ('PP', 149), ('Cs', 131), ('VO..."
1,022019041010400201001,01,04,04002,0400201001,Andalucía,Almería,Abrucena,1039,798,...,17841.0,17115.0,4640.0,4048.0,3418.0,2770.0,568.0,620.0,"{'PP': 127, 'PSOE': 380, 'Cs': 91, 'UP': 60, '...","[('PSOE', 380), ('PP', 127), ('VOX', 113), ('C..."
2,022019041010400301001,01,04,04003,0400301001,Andalucía,Almería,Adra,671,519,...,26498.0,24688.0,5121.0,4795.0,2499.0,2301.0,337.0,333.0,"{'PP': 162, 'PSOE': 131, 'Cs': 68, 'UP': 44, '...","[('PP', 162), ('PSOE', 131), ('VOX', 103), ('C..."
3,022019041010400301002,01,04,04003,0400301002,Andalucía,Almería,Adra,1282,954,...,25677.0,23400.0,5381.0,4837.0,1815.0,1724.0,343.0,464.0,"{'PP': 239, 'PSOE': 241, 'Cs': 166, 'UP': 62, ...","[('PSOE', 241), ('PP', 239), ('VOX', 218), ('C..."
4,022019041010400301003,01,04,04003,0400301003,Andalucía,Almería,Adra,1535,1087,...,22051.0,19687.0,5224.0,4044.0,1170.0,1198.0,416.0,476.0,"{'PP': 274, 'PSOE': 252, 'Cs': 170, 'UP': 67, ...","[('VOX', 282), ('PP', 274), ('PSOE', 252), ('C..."


Vemos que entre las columnas hay algunas como '45-50' que indica el número de personas entre 45 y 50 años que, creemos, no ofrece mucha imformación, en especial cuando tenemos otras columnas que segmentas por edades, como '%  mayores 65 años'.

In [ ]:
df_eleccion.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36317 entries, 0 to 36316
Data columns (total 97 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Sección                           36317 non-null  object 
 1   cod_ccaa                          36317 non-null  object 
 2   cod_prov                          36317 non-null  object 
 3   cod_mun                           36317 non-null  object 
 4   cod_sec                           36317 non-null  object 
 5   CCAA                              36317 non-null  object 
 6   Provincia                         36317 non-null  object 
 7   Municipio                         36317 non-null  object 
 8   Censo_Esc                         36317 non-null  int64  
 9   Votos_Total                       36317 non-null  int64  
 10  Participación                     36317 non-null  float64
 11  Nulos                             36317 non-null  int64  
 12  Voto

Ahora reunimos las columnas entre aquellas que son integer y las que son float, para hacer un formateado posterior.

In [ ]:
cols_integer = ['Censo_Esc', 'Votos_Total', 'Nulos', 'Votos_Válidos', 'Blanco', 'V_Cand', 'PP', 'PSOE', 'Cs', 'UP',
       'IU', 'VOX', 'UPyD', 'MP', 'CiU', 'ERC', 'JxC', 'CUP', 'DiL', 'PNV',
       'Bildu', 'Amaiur', 'CC', 'FA', 'TE', 'BNG', 'PRC', 'GBai', 'Compromis',
       'PACMA', 'Otros', 'Población Total', 'Hombres', 'Mujeres']

In [ ]:
cols_float = ['Participación', '% PP', '% PSOE', '% UP', '% Vox', '% Cs', '% IU',  '% mayores 65 años', '% 20-64 años', '% menores 19 años',
       'Afiliados SS Minicipio', '% Afiliados SS autónomos',
       '% Afiliados SS / Población', 'Paro Registrado Municipio',
       '% Paro Hombres', '% Paro mayores 45',
       '% Paro s/ Afiliados SS Municipio', 'Renta persona 2017',
       'Renta persona 2015', 'Renta hogar 2017', 'Renta hogar 2015',
       'Renta Salarios 2018', 'Renta Salarios 2015', 'Renta Pensiones 2018',
       'Renta Pensiones 2015', 'Renta Desempleo 2018', 'Renta Desempleo 2015']

Añadimos la columna que identifica la elección, en este caso 'Abril 2019'

In [ ]:
df_eleccion['Elecciones'] = 'Abril 2019'

Tenemos que formatear la columna de comunidad autónoma, ya que, increiblemente, el Ministerio del Interior y el INE tienen distintos códigos para ellas.

In [ ]:
ccaa = { 'Andalucía' : 'Andalucía',
        'Aragón' : 'Aragón',
         'Melilla' : 'Melilla',
         'Ceuta' : 'Ceuta',
        'Cantabria' : 'Cantabria',
        'Asturias' : 'Asturias',
        "'Islas Baleares'": 'Islas Baleares',
         'Cataluña' : 'Cataluña',
        'Galicia' : 'Madrid',
        'Extremadura' : 'Galicia',
        'Castilla y León' : 'Castilla - La Mancha',
        'Castilla - La Mancha' : 'Castilla y León',
        'C. Valenciana' : 'Extremadura',
        'La Rioja' : 'C. Valenciana',
        'País Vasco' : 'La Rioja',
        'Murcia' : 'País Vasco',
        'Navarra' : 'Murcia',
        'Madrid' : 'Navarra',
        'Canarias' : 'Islas Canarias'
}

Mapeamos la columna de comunidad autónomas.

In [ ]:
df_eleccion['CCAA'] = df_eleccion['CCAA'].map(ccaa)

Definimos la columna 'Distrito' por si fuese útil en el futuro.

In [ ]:
df_eleccion['Distrito'] = df_eleccion['Sección'].str[16:18]

Seleccionamos las columnas que queremos, eliminando las de edades con intervalos de 5 años.

In [ ]:
df_eleccion = df_eleccion[['Elecciones', 'Sección', 'cod_ccaa', 'cod_prov', 'cod_mun', 'cod_sec', 'CCAA',
       'Provincia', 'Municipio', 'Distrito', 'Censo_Esc', 'Votos_Total', 'Participación',
       'Nulos', 'Votos_Válidos', 'Blanco', 'V_Cand', 'PP', 'PSOE', 'Cs', 'UP',
       'IU', 'VOX', 'UPyD', 'MP', 'CiU', 'ERC', 'JxC', 'CUP', 'DiL', 'PNV',
       'Bildu', 'Amaiur', 'CC', 'FA', 'TE', 'BNG', 'PRC', 'GBai', 'Compromis',
       'PACMA', 'Otros', '% PP', '% PSOE', '% UP', '% Vox', '% Cs', '% IU',
       'Ganador', 'Segundo', 'Tercero', 'Cuarto', 'Quinto', 'Población Total', 'Hombres', 'Mujeres',
       '% mayores 65 años', '% 20-64 años', '% menores 19 años',
       'Afiliados SS Minicipio', '% Afiliados SS autónomos',
       '% Afiliados SS / Población', 'Paro Registrado Municipio',
       '% Paro Hombres', '% Paro mayores 45',
       '% Paro s/ Afiliados SS Municipio', 'Renta persona 2017',
       'Renta persona 2015', 'Renta hogar 2017', 'Renta hogar 2015',
       'Renta Salarios 2018', 'Renta Salarios 2015', 'Renta Pensiones 2018',
       'Renta Pensiones 2015', 'Renta Desempleo 2018', 'Renta Desempleo 2015']]

Seleccionamos las columnas para eliminar los na.

In [ ]:
cols_na = ['Censo_Esc',
       'Votos_Total', 'Participación', 'Nulos', 'Votos_Válidos', 'Blanco',
       'V_Cand', 'PP', 'PSOE', 'Cs', 'UP', 'IU', 'VOX', 'UPyD', 'MP', 'CiU',
       'ERC', 'JxC', 'CUP', 'DiL', 'PNV', 'Bildu', 'Amaiur', 'CC', 'FA', 'TE',
       'BNG', 'PRC', 'GBai', 'Compromis', 'PACMA', 'Otros', '% PP', '% PSOE',
       '% UP', '% Vox', '% Cs', '% IU', 'Población Total', 'Hombres', 'Mujeres',
       '% mayores 65 años', '% 20-64 años', '% menores 19 años',
       'Afiliados SS Minicipio', '% Afiliados SS autónomos',
       '% Afiliados SS / Población', 'Paro Registrado Municipio',
       '% Paro Hombres', '% Paro mayores 45',
       '% Paro s/ Afiliados SS Municipio', 'Renta persona 2017',
       'Renta persona 2015', 'Renta hogar 2017', 'Renta hogar 2015',
       'Renta Salarios 2018', 'Renta Salarios 2015', 'Renta Pensiones 2018',
       'Renta Pensiones 2015', 'Renta Desempleo 2018', 'Renta Desempleo 2015']

Cambiamos el nombre de la columna '% Vox', para evitar problemas en el futuro.

In [ ]:
df_eleccion = df_eleccion.rename(columns = {'% Vox' : '% VOX'})

Ahora almacenamos todos estos cambios en una sola función, que aplicaremos a los 5 datasets, incluyendo un formato inicial a la columna del nombre de municipio.

In [ ]:
def transformacion(df_eleccion, fecha):

  df_eleccion = df_eleccion[['Sección', 'cod_ccaa', 'cod_prov', 'cod_mun', 'cod_sec', 'CCAA',
       'Provincia', 'Municipio', 'Censo_Esc', 'Votos_Total', 'Participación',
       'Nulos', 'Votos_Válidos', 'Blanco', 'V_Cand', 'PP', 'PSOE', 'Cs', 'UP',
       'IU', 'VOX', 'UPyD', 'MP', 'CiU', 'ERC', 'JxC', 'CUP', 'DiL', 'PNV',
       'Bildu', 'Amaiur', 'CC', 'FA', 'TE', 'BNG', 'PRC', 'GBai', 'Compromis',
       'PACMA', 'Otros', '% PP', '% PSOE', '% UP', '% Vox', '% Cs', '% IU',
       'Ganador', 'Segundo', 'Tercero', 'Cuarto', 'Quinto', 'Población Total', 'Hombres', 'Mujeres',
       '% mayores 65 años', '% 20-64 años', '% menores 19 años',
       'Afiliados SS Minicipio', '% Afiliados SS autónomos',
       '% Afiliados SS / Población', 'Paro Registrado Municipio',
       '% Paro Hombres', '% Paro mayores 45',
       '% Paro s/ Afiliados SS Municipio', 'Renta persona 2017',
       'Renta persona 2015', 'Renta hogar 2017', 'Renta hogar 2015',
       'Renta Salarios 2018', 'Renta Salarios 2015', 'Renta Pensiones 2018',
       'Renta Pensiones 2015', 'Renta Desempleo 2018', 'Renta Desempleo 2015']]

  cols_integer = ['Censo_Esc', 'Votos_Total', 'Nulos', 'Votos_Válidos', 'Blanco', 'V_Cand', 'PP', 'PSOE', 'Cs', 'UP',
       'IU', 'VOX', 'UPyD', 'MP', 'CiU', 'ERC', 'JxC', 'CUP', 'DiL', 'PNV',
       'Bildu', 'Amaiur', 'CC', 'FA', 'TE', 'BNG', 'PRC', 'GBai', 'Compromis',
       'PACMA', 'Otros', 'Población Total', 'Hombres', 'Mujeres']

  cols_float = ['Participación', '% PP', '% PSOE', '% UP', '% Vox', '% Cs', '% IU',  '% mayores 65 años', '% 20-64 años', '% menores 19 años',
       'Afiliados SS Minicipio', '% Afiliados SS autónomos',
       '% Afiliados SS / Población', 'Paro Registrado Municipio',
       '% Paro Hombres', '% Paro mayores 45',
       '% Paro s/ Afiliados SS Municipio', 'Renta persona 2017',
       'Renta persona 2015', 'Renta hogar 2017', 'Renta hogar 2015',
       'Renta Salarios 2018', 'Renta Salarios 2015', 'Renta Pensiones 2018',
       'Renta Pensiones 2015', 'Renta Desempleo 2018', 'Renta Desempleo 2015']

  # List of columns to fill NaN values, defined based on original notebook's cols_na
  cols_to_fill_na = ['Censo_Esc', 'Votos_Total', 'Participación', 'Nulos', 'Votos_Válidos', 'Blanco',
       'V_Cand', 'PP', 'PSOE', 'Cs', 'UP', 'IU', 'VOX', 'UPyD', 'MP', 'CiU',
       'ERC', 'JxC', 'CUP', 'DiL', 'PNV', 'Bildu', 'Amaiur', 'CC', 'FA', 'TE',
       'BNG', 'PRC', 'GBai', 'Compromis', 'PACMA', 'Otros', '% PP', '% PSOE',
       '% UP', '% Vox', '% Cs', '% IU', 'Población Total', 'Hombres', 'Mujeres',
       '% mayores 65 años', '% 20-64 años', '% menores 19 años',
       'Afiliados SS Minicipio', '% Afiliados SS autónomos',
       '% Afiliados SS / Población', 'Paro Registrado Municipio',
       '% Paro Hombres', '% Paro mayores 45',
       '% Paro s/ Afiliados SS Municipio', 'Renta persona 2017',
       'Renta persona 2015', 'Renta hogar 2017', 'Renta hogar 2015',
       'Renta Salarios 2018', 'Renta Salarios 2015', 'Renta Pensiones 2018',
       'Renta Pensiones 2015', 'Renta Desempleo 2018', 'Renta Desempleo 2015']

  # Fill NaN values for all relevant numeric columns first
  for col in cols_to_fill_na:
    if col in df_eleccion.columns:
      df_eleccion[col] = df_eleccion[col].fillna(df_eleccion[col].mean(numeric_only=True))

  # Now, safely convert integer columns
  for col in cols_integer:
    df_eleccion[col] = df_eleccion[col].astype('int32')

  # And convert float columns
  for col in cols_float:
    df_eleccion[col] = df_eleccion[col].astype('float32')

  df_eleccion['Elecciones'] = fecha

  ccaa = { 'Andalucía' : 'Andalucía',
        'Aragón' : 'Aragón',
         'Melilla' : 'Melilla',
         'Ceuta' : 'Ceuta',
        'Cantabria' : 'Cantabria',
        'Asturias' : 'Asturias',
        "'Islas Baleares'": 'Islas Baleares',
         'Cataluña' : 'Cataluña',
        'Galicia' : 'Madrid',
        'Extremadura' : 'Galicia',
        'Castilla y León' : 'Castilla - La Mancha',
        'Castilla - La Mancha' : 'Castilla y León',
        'C. Valenciana' : 'Extremadura',
        'La Rioja' : 'C. Valenciana',
        'País Vasco' : 'La Rioja',
        'Murcia' : 'País Vasco',
        'Navarra' : 'Murcia',
        'Madrid' : 'Navarra',
        'Canarias' : 'Islas Canarias'
}
  df_eleccion['CCAA'] = df_eleccion['CCAA'].map(ccaa)

  df_eleccion['Distrito'] = df_eleccion['Sección'].str[16:18]

  # Reorder columns to include newly created ones
  final_cols_order = ['Elecciones', 'Sección', 'cod_ccaa', 'cod_prov', 'cod_mun', 'cod_sec', 'CCAA',
       'Provincia', 'Municipio', 'Distrito', 'Censo_Esc', 'Votos_Total', 'Participación',
       'Nulos', 'Votos_Válidos', 'Blanco', 'V_Cand', 'PP', 'PSOE', 'Cs', 'UP',
       'IU', 'VOX', 'UPyD', 'MP', 'CiU', 'ERC', 'JxC', 'CUP', 'DiL', 'PNV',
       'Bildu', 'Amaiur', 'CC', 'FA', 'TE', 'BNG', 'PRC', 'GBai', 'Compromis',
       'PACMA', 'Otros', '% PP', '% PSOE', '% UP', '% Vox', '% Cs', '% IU',
       'Ganador', 'Segundo', 'Tercero', 'Cuarto', 'Quinto', 'Población Total', 'Hombres', 'Mujeres',
       '% mayores 65 años', '% 20-64 años', '% menores 19 años',
       'Afiliados SS Minicipio', '% Afiliados SS autónomos',
       '% Afiliados SS / Población', 'Paro Registrado Municipio',
       '% Paro Hombres', '% Paro mayores 45',
       '% Paro s/ Afiliados SS Municipio', 'Renta persona 2017',
       'Renta persona 2015', 'Renta hogar 2017', 'Renta hogar 2015',
       'Renta Salarios 2018', 'Renta Salarios 2015', 'Renta Pensiones 2018',
       'Renta Pensiones 2015', 'Renta Desempleo 2018', 'Renta Desempleo 2015']

  df_eleccion = df_eleccion[final_cols_order]

  df_eleccion = df_eleccion.rename(columns = {'% Vox' : '% VOX'})

  df_eleccion['Municipio'] = df_eleccion['Municipio'].str.title()

  return df_eleccion

Ahora aplicamos la función desde cero al dataset de Abril 2019, generado, por desgracia, unos warnings, lo sentimos.

In [ ]:
df_eleccion_A19 = pd.read_csv('/content/drive/MyDrive/Proyecto_KeepCoding - Propio/Data/Gen-19-Abr/gen_A19_unif_cols_prov.txt', dtype = strings)

In [ ]:
df_eleccion_A19 = transformacion(df_eleccion_A19, 'Abril 2019')

/tmp/ipython-input-4283326090.py:51: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eleccion[col] = df_eleccion[col].fillna(df_eleccion[col].mean(numeric_only=True))
/tmp/ipython-input-4283326090.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eleccion[col] = df_eleccion[col].astype('int32')
/tmp/ipython-input-4283326090.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the 

In [ ]:
df_eleccion_A19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36317 entries, 0 to 36316
Data columns (total 76 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Elecciones                        36317 non-null  object 
 1   Sección                           36317 non-null  object 
 2   cod_ccaa                          36317 non-null  object 
 3   cod_prov                          36317 non-null  object 
 4   cod_mun                           36317 non-null  object 
 5   cod_sec                           36317 non-null  object 
 6   CCAA                              36317 non-null  object 
 7   Provincia                         36317 non-null  object 
 8   Municipio                         36317 non-null  object 
 9   Distrito                          36317 non-null  object 
 10  Censo_Esc                         36317 non-null  int32  
 11  Votos_Total                       36317 non-null  int32  
 12  Part

In [ ]:
df_eleccion_A19.describe()

,Censo_Esc,Votos_Total,Participación,Nulos,Votos_Válidos,Blanco,V_Cand,PP,PSOE,Cs,...,Renta persona 2017,Renta persona 2015,Renta hogar 2017,Renta hogar 2015,Renta Salarios 2018,Renta Salarios 2015,Renta Pensiones 2018,Renta Pensiones 2015,Renta Desempleo 2018,Renta Desempleo 2015
count,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,...,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000,36317.000000
mean,958.261117,725.824903,0.758828,7.581408,718.243495,5.482749,712.760746,122.912851,206.010050,113.907674,...,11505.018555,10869.106445,29477.535156,28020.873047,8314.099609,7306.835938,3178.389648,2940.962891,289.498322,346.348419
std,437.906979,344.834412,0.073585,6.115915,341.724321,4.412176,339.188435,87.894068,112.301806,82.379609,...,3643.010986,3524.067627,10072.783203,9680.139648,3463.090332,3275.538574,1356.940308,1210.196045,115.998802,114.511986
min,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,3431.000000,3281.000000,12087.000000,11388.000000,1767.000000,1349.000000,345.000000,476.000000,71.000000,81.000000
25%,691.000000,509.000000,0.720143,3.000000,503.000000,2.000000,499.000000,54.000000,132.000000,55.000000,...,9109.000000,8520.000000,23242.000000,21935.000000,6084.000000,5160.000000,2269.000000,2123.000000,219.000000,279.000000
50%,952.000000,713.000000,0.766423,6.000000,706.000000,5.000000,701.000000,108.000000,198.000000,101.000000,...,10936.000000,10304.000000,27620.000000,26185.000000,7855.336426,6845.948242,3112.000000,2894.000000,276.000000,343.000000
75%,1242.000000,945.000000,0.807858,11.000000,936.000000,8.000000,929.000000,171.000000,274.000000,157.000000,...,12934.000000,12291.000000,32871.000000,31388.000000,9593.000000,8561.000000,3809.000000,3511.000000,322.000000,388.000000
max,3845.000000,3075.000000,1.000000,77.000000,3058.000000,168.000000,3042.000000,1016.000000,799.000000,725.000000,...,30210.000000,28350.000000,86006.000000,80135.000000,26674.000000,24252.000000,8302.000000,7398.000000,846.000000,877.000000


Aplicamos ahora la función al dataset de Noviembre 2019.

In [ ]:
df_eleccion_Nov19 = pd.read_csv('/content/drive/MyDrive/Proyecto_KeepCoding - Propio/Data/Gen-19-Nov/gen_N19_unif_cols_prov.txt', dtype = strings)

In [ ]:
df_eleccion_N19 = transformacion(df_eleccion_Nov19, 'Noviembre 2019')

/tmp/ipython-input-4283326090.py:51: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eleccion[col] = df_eleccion[col].fillna(df_eleccion[col].mean(numeric_only=True))
/tmp/ipython-input-4283326090.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eleccion[col] = df_eleccion[col].astype('int32')
/tmp/ipython-input-4283326090.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the 

In [ ]:
df_eleccion_N19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36302 entries, 0 to 36301
Data columns (total 76 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Elecciones                        36302 non-null  object 
 1   Sección                           36302 non-null  object 
 2   cod_ccaa                          36302 non-null  object 
 3   cod_prov                          36302 non-null  object 
 4   cod_mun                           36302 non-null  object 
 5   cod_sec                           36302 non-null  object 
 6   CCAA                              36302 non-null  object 
 7   Provincia                         36302 non-null  object 
 8   Municipio                         36302 non-null  object 
 9   Distrito                          36302 non-null  object 
 10  Censo_Esc                         36302 non-null  int32  
 11  Votos_Total                       36302 non-null  int32  
 12  Part

Ahora lo aplicamos al dataset de Noviembre 2011.

In [ ]:
df_eleccion_11 = pd.read_csv('/content/drive/MyDrive/Proyecto_KeepCoding - Propio/Data/Gen-11-Nov/gen_N11_unif_cols_prov.txt', dtype = strings)

In [ ]:
df_eleccion_N11 = transformacion(df_eleccion_11, 'Noviembre 2011')

/tmp/ipython-input-4283326090.py:51: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eleccion[col] = df_eleccion[col].fillna(df_eleccion[col].mean(numeric_only=True))
/tmp/ipython-input-4283326090.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eleccion[col] = df_eleccion[col].astype('int32')
/tmp/ipython-input-4283326090.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the 

In [ ]:
df_eleccion_N11.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36063 entries, 0 to 36062
Data columns (total 76 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Elecciones                        36063 non-null  object 
 1   Sección                           36063 non-null  object 
 2   cod_ccaa                          36063 non-null  object 
 3   cod_prov                          36063 non-null  object 
 4   cod_mun                           36063 non-null  object 
 5   cod_sec                           36063 non-null  object 
 6   CCAA                              36063 non-null  object 
 7   Provincia                         36063 non-null  object 
 8   Municipio                         36063 non-null  object 
 9   Distrito                          36063 non-null  object 
 10  Censo_Esc                         36063 non-null  int32  
 11  Votos_Total                       36063 non-null  int32  
 12  Part

Ahora lo aplicamos al dataset de diciembre 2015.

In [ ]:
df_eleccion_15 = pd.read_csv('/content/drive/MyDrive/Proyecto_KeepCoding - Propio/Data/Gen-15-Dic/gen_D15_unif_cols_prov.txt', dtype = strings)

In [ ]:
df_eleccion_D15 = transformacion(df_eleccion_15, 'Diciembre 2015')

/tmp/ipython-input-4283326090.py:51: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eleccion[col] = df_eleccion[col].fillna(df_eleccion[col].mean(numeric_only=True))
/tmp/ipython-input-4283326090.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eleccion[col] = df_eleccion[col].astype('int32')
/tmp/ipython-input-4283326090.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the 

In [ ]:
df_eleccion_D15.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36217 entries, 0 to 36216
Data columns (total 76 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Elecciones                        36217 non-null  object 
 1   Sección                           36217 non-null  object 
 2   cod_ccaa                          36217 non-null  object 
 3   cod_prov                          36217 non-null  object 
 4   cod_mun                           36217 non-null  object 
 5   cod_sec                           36217 non-null  object 
 6   CCAA                              36217 non-null  object 
 7   Provincia                         36217 non-null  object 
 8   Municipio                         36217 non-null  object 
 9   Distrito                          36217 non-null  object 
 10  Censo_Esc                         36217 non-null  int32  
 11  Votos_Total                       36217 non-null  int32  
 12  Part

Y, finalmente, aplicamos la función al dataset de Junio 2016.

In [ ]:
df_eleccion_16 = pd.read_csv('/content/drive/MyDrive/Proyecto_KeepCoding - Propio/Data/Gen-16-Jun/gen_J16_unif_cols_prov.txt', dtype = strings)

In [ ]:
df_eleccion_J16 = transformacion(df_eleccion_16, 'Junio 2016')

/tmp/ipython-input-4283326090.py:51: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eleccion[col] = df_eleccion[col].fillna(df_eleccion[col].mean(numeric_only=True))
/tmp/ipython-input-4283326090.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eleccion[col] = df_eleccion[col].astype('int32')
/tmp/ipython-input-4283326090.py:55: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the 

In [ ]:
df_eleccion_J16.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36193 entries, 0 to 36192
Data columns (total 76 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Elecciones                        36193 non-null  object 
 1   Sección                           36193 non-null  object 
 2   cod_ccaa                          36193 non-null  object 
 3   cod_prov                          36193 non-null  object 
 4   cod_mun                           36193 non-null  object 
 5   cod_sec                           36193 non-null  object 
 6   CCAA                              36193 non-null  object 
 7   Provincia                         36193 non-null  object 
 8   Municipio                         36193 non-null  object 
 9   Distrito                          36193 non-null  object 
 10  Censo_Esc                         36193 non-null  int32  
 11  Votos_Total                       36193 non-null  int32  
 12  Part

Ahora hacemos un concat a los 5 dataset, y tenemos un dataset unificado de más de 181.000 filas.

Vemos que, de momento, tendremos tipados de int32 y float32, aunque finalmente pandas en el futuro lo cargará en int64 y float64.

In [ ]:
df_eleccion_unificado = pd.concat([df_eleccion_A19, df_eleccion_N19, df_eleccion_N11, df_eleccion_D15, df_eleccion_J16], ignore_index=True)

In [ ]:
df_eleccion_unificado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181092 entries, 0 to 181091
Data columns (total 76 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   Elecciones                        181092 non-null  object 
 1   Sección                           181092 non-null  object 
 2   cod_ccaa                          181092 non-null  object 
 3   cod_prov                          181092 non-null  object 
 4   cod_mun                           181092 non-null  object 
 5   cod_sec                           181092 non-null  object 
 6   CCAA                              181092 non-null  object 
 7   Provincia                         181092 non-null  object 
 8   Municipio                         181092 non-null  object 
 9   Distrito                          181092 non-null  object 
 10  Censo_Esc                         181092 non-null  int32  
 11  Votos_Total                       181092 non-null  i

In [ ]:
df_eleccion_unificado.head()

,Elecciones,Sección,cod_ccaa,cod_prov,cod_mun,cod_sec,CCAA,Provincia,Municipio,Distrito,...,Renta persona 2017,Renta persona 2015,Renta hogar 2017,Renta hogar 2015,Renta Salarios 2018,Renta Salarios 2015,Renta Pensiones 2018,Renta Pensiones 2015,Renta Desempleo 2018,Renta Desempleo 2015
0,Abril 2019,022019041010400101001,01,04,04001,0400101001,Andalucía,Almería,Abla,01,...,9159.0,8788.0,20172.0,19546.0,5574.0,4833.0,3286.0,3082.0,403.0,471.0
1,Abril 2019,022019041010400201001,01,04,04002,0400201001,Andalucía,Almería,Abrucena,01,...,8827.0,8107.0,17841.0,17115.0,4640.0,4048.0,3418.0,2770.0,568.0,620.0
2,Abril 2019,022019041010400301001,01,04,04003,0400301001,Andalucía,Almería,Adra,01,...,8965.0,8267.0,26498.0,24688.0,5121.0,4795.0,2499.0,2301.0,337.0,333.0
3,Abril 2019,022019041010400301002,01,04,04003,0400301002,Andalucía,Almería,Adra,01,...,8599.0,7941.0,25677.0,23400.0,5381.0,4837.0,1815.0,1724.0,343.0,464.0
4,Abril 2019,022019041010400301003,01,04,04003,0400301003,Andalucía,Almería,Adra,01,...,8076.0,7150.0,22051.0,19687.0,5224.0,4044.0,1170.0,1198.0,416.0,476.0


In [ ]:
df_eleccion_unificado.describe()

,Censo_Esc,Votos_Total,Participación,Nulos,Votos_Válidos,Blanco,V_Cand,PP,PSOE,Cs,...,Renta persona 2017,Renta persona 2015,Renta hogar 2017,Renta hogar 2015,Renta Salarios 2018,Renta Salarios 2015,Renta Pensiones 2018,Renta Pensiones 2015,Renta Desempleo 2018,Renta Desempleo 2015
count,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,...,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000,181092.000000
mean,957.034828,689.493489,0.723158,7.141470,682.352020,6.160725,676.191295,196.553023,177.728149,68.460368,...,11496.067383,10862.742188,29433.285156,27989.615234,8282.939453,7284.106934,3192.533936,2951.170654,289.504608,346.319580
std,436.514202,326.914110,0.080056,5.879054,323.820518,5.141824,320.933886,148.265702,106.550202,72.613028,...,3634.063477,3523.936035,10028.007812,9675.264648,3428.995361,3264.528809,1350.194946,1209.791870,115.795799,114.663216
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,3431.000000,3281.000000,12087.000000,11388.000000,1767.000000,1349.000000,345.000000,476.000000,71.000000,81.000000
25%,689.000000,483.000000,0.677089,3.000000,478.000000,2.000000,474.000000,84.000000,104.000000,6.000000,...,9107.000000,8513.000000,23229.000000,21912.000000,6077.000000,5146.000000,2287.000000,2135.000000,219.000000,279.000000
50%,951.000000,677.000000,0.729825,6.000000,670.000000,5.000000,664.000000,164.000000,166.000000,51.000000,...,10931.000000,10296.000000,27598.000000,26157.000000,7855.336426,6845.948242,3130.000000,2908.000000,276.000000,343.000000
75%,1239.000000,898.000000,0.777854,10.000000,889.000000,9.000000,881.000000,273.000000,238.000000,103.000000,...,12924.000000,12282.000000,32820.000000,31348.000000,9547.000000,8527.000000,3817.000000,3517.000000,322.000000,387.000000
max,6670.000000,4852.000000,1.000000,165.000000,4794.000000,168.000000,4693.000000,1322.000000,1729.000000,841.000000,...,30210.000000,28350.000000,86006.000000,80135.000000,26674.000000,24252.000000,8302.000000,7398.000000,846.000000,877.000000


No hemos terminado. Tenemos un problema con los nombres de los municipios, ya que en la elección de 2015 algunos de ellos eran diferentes a los del resto de dataset.

Por ejemplo, en la provincia de Madrid, tenemos un municipio de 'Los Molinos' y 'Molinos (Los)', que hacen referencia al mismo.

In [ ]:
muni = df_eleccion_unificado.loc[df_eleccion_unificado['Provincia'] == 'Madrid']['Municipio'].unique()

In [ ]:
mun = df_eleccion_unificado['Municipio'].unique()

In [ ]:
len(muni)

193

In [ ]:
muni

array(['La Acebeda', 'Ajalvir', 'Alameda Del Valle', 'El Álamo',
       'Alcalá De Henares', 'Alcobendas', 'Alcorcón', 'Aldea Del Fresno',
       'Algete', 'Alpedrete', 'Ambite', 'Anchuelo', 'Aranjuez',
       'Arganda Del Rey', 'Arroyomolinos', 'El Atazar', 'Batres',
       'Becerril De La Sierra', 'Belmonte De Tajo', 'Berzosa Del Lozoya',
       'El Berrueco', 'Boadilla Del Monte', 'El Boalo', 'Braojos',
       'Brea De Tajo', 'Brunete', 'Buitrago Del Lozoya', 'Bustarviejo',
       'Cabanillas De La Sierra', 'La Cabrera', 'Cadalso De Los Vidrios',
       'Camarma De Esteruelas', 'Campo Real', 'Canencia', 'Carabaña',
       'Casarrubuelos', 'Cenicientos', 'Cercedilla',
       'Cervera De Buitrago', 'Ciempozuelos', 'Cobeña',
       'Colmenar Del Arroyo', 'Colmenar De Oreja', 'Colmenarejo',
       'Colmenar Viejo', 'Collado Mediano', 'Collado Villalba', 'Corpa',
       'Coslada', 'Cubas De La Sagra', 'Chapinería', 'Chinchón',
       'Daganzo De Arriba', 'El Escorial', 'Estremera',
     

Afortunadamente, el código de cada municipio no varía, por lo que podemos hacer un diccionario con este código y uno de los nombres, en caso de que haya dos.

In [ ]:
dict_cod_mun = {}

In [ ]:
for x, y in zip(df_eleccion_unificado['cod_mun'], df_eleccion_unificado['Municipio']):
  dict_cod_mun[x] = y

In [ ]:
len(dict_cod_mun)

8135

In [ ]:
df_eleccion_unificado['Municipio'] = df_eleccion_unificado['cod_mun'].map(dict_cod_mun)

Ahora hacemos un mapeo, y encontramos para toda España un número de municipios de 8.118, que es el correcto.

In [ ]:
len(df_eleccion_unificado['Municipio'].unique())

8118

In [ ]:
muni_can = df_eleccion_unificado.loc[df_eleccion_unificado['Provincia'] == 'Cantabria']['Municipio'].unique()

Ya tenemos 102 municipios en Cantabria, el número correcto.

In [ ]:
len(muni_can)

102

Antes teníamos 193 municipios en Madrid, y ahora 179, que es el número correcto.

In [ ]:
muni_mad = df_eleccion_unificado.loc[df_eleccion_unificado['Provincia'] == 'Madrid']['Municipio'].unique()

In [ ]:
len(muni_mad)

179

In [ ]:
muni_mad

array(['La Acebeda', 'Ajalvir', 'Alameda Del Valle', 'El Álamo',
       'Alcalá De Henares', 'Alcobendas', 'Alcorcón', 'Aldea Del Fresno',
       'Algete', 'Alpedrete', 'Ambite', 'Anchuelo', 'Aranjuez',
       'Arganda Del Rey', 'Arroyomolinos', 'El Atazar', 'Batres',
       'Becerril De La Sierra', 'Belmonte De Tajo', 'Berzosa Del Lozoya',
       'El Berrueco', 'Boadilla Del Monte', 'El Boalo', 'Braojos',
       'Brea De Tajo', 'Brunete', 'Buitrago Del Lozoya', 'Bustarviejo',
       'Cabanillas De La Sierra', 'La Cabrera', 'Cadalso De Los Vidrios',
       'Camarma De Esteruelas', 'Campo Real', 'Canencia', 'Carabaña',
       'Casarrubuelos', 'Cenicientos', 'Cercedilla',
       'Cervera De Buitrago', 'Ciempozuelos', 'Cobeña',
       'Colmenar Del Arroyo', 'Colmenar De Oreja', 'Colmenarejo',
       'Colmenar Viejo', 'Collado Mediano', 'Collado Villalba', 'Corpa',
       'Coslada', 'Cubas De La Sagra', 'Chapinería', 'Chinchón',
       'Daganzo De Arriba', 'El Escorial', 'Estremera',
     

In [ ]:
df_eleccion_unificado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181092 entries, 0 to 181091
Data columns (total 76 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   Elecciones                        181092 non-null  object 
 1   Sección                           181092 non-null  object 
 2   cod_ccaa                          181092 non-null  object 
 3   cod_prov                          181092 non-null  object 
 4   cod_mun                           181092 non-null  object 
 5   cod_sec                           181092 non-null  object 
 6   CCAA                              181092 non-null  object 
 7   Provincia                         181092 non-null  object 
 8   Municipio                         181092 non-null  object 
 9   Distrito                          181092 non-null  object 
 10  Censo_Esc                         181092 non-null  int32  
 11  Votos_Total                       181092 non-null  i

Ya podemos guardar pues el dataset unificado y curado.

In [ ]:
df_eleccion_unificado.to_csv('/content/drive/MyDrive/Practica_Bootcamp_IA_4/eleccion_unificado.csv', index = False)